Auto clear primary clipboard

codokie opened the pull request:
I modified the clipboard history manager so that it automatically clears clipboard history items which are older than the retention time (right now items are cleared just before the clipboard history view is shown).

I also added an option in Preferences to auto clear the primary clipboard as well as the history ~~(only if it wasn't pinned)~~.
I believe this enhances the privacy of those who need to copy sensitive information like passwords (Android 13+ clears the clipboard only after an hour).

Additionally, I added the 'clear clipboard history' button as toolbar utility button (for some unclear reason it has been filtered out).

Lastly, I ensured that if you copied text already present in history, then the old entry would have its timestamp and position updated instead of adding a new (duplicated) entry.

codokie commented:
Regarding #306, in [e385068](https://github.com/Helium314/HeliBoard/pull/611/commits/e38506888979a2946828646b3e4ea9430c8a1e55)
I added a "Copy to internal clipboard" toggle in Preferences with summary:

> If disabled, copy key will copy to the system primary clipboard as well

It can be activated only if the clipboard history is enabled.

Helium314 commented:
I didn't have much time and will not have time for this at least for another day. I just want to say that the merge conflict is coming from a quick and dirty maybe-fix for the duplicate clipboard entries. There is no reason to keep this fix of mine, it was just for the necessary 1.2 release.

codokie commented:
It shouldn't be a problem. I added a new function `copyTextToInternalClipboard` in [e385068](https://github.com/Helium314/HeliBoard/pull/611/commits/e38506888979a2946828646b3e4ea9430c8a1e55) that makes sure to update the timestamp and position of entries that are copied twice instead of adding a duplicate.

I've also added a cut key which I believe is very useful to have.
Those who would like to test the new keys should note that they are not added to the toolbar by default.
In order to see them you have to click "Default" in the reorder popup key dialog because upgradeToolbarPref() is called only after a new version of the app has been installed.

codokie commented:
@Helium314 I guess I got carried away.. Sorry.
I'll do my best to add the clipboard suggestions as a new PR, and I'll make sure not to use this PR for other related new features.
I hope it's okay to leave the already committed additions here.

Helium314 commented:
> I modified the clipboard history manager so that it automatically clears clipboard history items which are older than the retention time (right now items are cleared just before the clipboard history view is shown).

Why does this matter?

> I also added an option in Preferences to auto clear the primary clipboard as well as the history (only if it wasn't pinned).
I believe this enhances the privacy of those who need to copy sensitive information like passwords (Android 13+ clears the clipboard only after an hour).

I'm not sure here. I see that it could be useful, but at the same time this is functionality that seems rather unrelated to a keyboard.
Has this been requested anywhere?

> Additionally, I added the 'clear clipboard history' button as toolbar utility button (for some unclear reason it has been filtered out).

The reason is that you do not see the clipboard from normal tool bar. If the reason is unclear, you can just ask.

> I added a "Copy to internal clipboard" toggle in Preferences with summary

As mentioned in that issue, I think a separate key would be preferable.
Or is there a reason why you think the setting is better?

> because upgradeToolbarPref() is called only after a new version of the app has been installed

Hmm, I wonder whether we could just call it on every start for debug builds...

codokie commented:
> Why does this matter?

Because if you long press the clipboard key, it will paste the contents of the last entry even after retention time.
>Has this been requested anywhere?

Probably not, but I find it very useful and it seems fitting for a privacy focused keyboard.
I presume you can achieve that with Tasker? I haven't figured out how to do it yet.
I know this app is a keyboard app, but since it has a clipboard manager, that should offer some customizability too.
>The reason is that you do not see the clipboard from normal tool bar

I personally feel uneasy to go to the clipboard history view just to press that button when I copy a password because it is exposed there, so I don't agree with the button being filtered out.

> Or is there a reason why you think the setting is better?

A button would be good as well but I am not sure about the icon of it, it needs to be distinguishable from the regular one.
I also have not seen a keyboard with such a button yet, and the new cut key will need a counterpart as well.
Those who enable the setting can still copy text to the primary clipboard by other means.

Helium314 commented:
> I think it is better to get rid of those "internal clipboard" entries ASAP and not only when the user is about to open a view.

If anything that is not the app can read it, then it likely already has root access and know exactly where to look, so we don't need to bother anyway.

> A button would be good as well but I am not sure about the icon of it, it needs to be distinguishable from the regular one.

Maybe the icon could be some mashup between copy and incognito icons?

------

I have the feeling for some changes there is more discussion necessary, and I see how annoyed I am by looking at that bunch of unrelated changes (and I feel is biases my judgement).
So if it's not a huge amount of work for you, could you please split the changes into separate PRs? (There is a reason size is in many [good-PR](https://swiftpublished.com/article/how-to-create-a-good-pull-request) [guides](https://developers.google.com/blockly/guides/contribute/get-started/write_a_good_pr))

Then the obviously ok parts can be merged right away, and do not stay permanently in view when I'm looking at other stuff.
Also with changes I'm critical about it might help when I straightforward see the changes without having to pick them out from everything else.

codokie commented:
@Helium314 Could you tell me what parts are **not** problematic so I will open a separate PR for them?
Maybe we can keep the discussion here until all features are merged and/or dismissed.
> If anything that is not the app can read it, then it likely already has root access 

Think about this scenario: you try to sign in to some website in your browser and for some reason the autofill service isn't working, so you go to your password manager to copy and paste your password. You close the browser, and proceed to an app which then pastes from your clipboard.
That app just got your password to some potentially important website without root access. In older Android versions you won't even know this happened.
> Maybe the icon could be some mashup between copy and incognito icons?
That could be done but I kinda think the toolbar list is somewhat crowded already.

Florisboard has the same setting as the proposed one but with a different naming ("Sync to system clipboard" toggled off), so I figured we better not try to reinvent the wheel.

codokie commented:
@Helium314  I am closing this PR and will open a new one shortly only for the "auto clear primary clipboard" functionality. thanks
